File: backend/app/solver/domain.py
"""
Compilación del problema: del JSON de entrada a estructuras indexadas por entero.

El solver nunca vuelve a tocar los modelos Pydantic. Todo se traduce una sola vez a
listas y sets indexados por posición, porque el backtracking hace millones de
comprobaciones de pertenencia y `dict[str]` en el hot loop cuesta caro.

Vocabulario:
  slot   = par (día, bloque de clase) aplanado a un entero 0..S-1
  lesson = una hora suelta de (grupo, materia). Un plan con 5 h de Matemáticas
           en 1°A genera 5 lessons independientes. Son las variables del CSP.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Dict, List, Optional, Sequence, Set, Tuple

from ..models.schemas import (
    AvailabilityMode,
    Group,
    ScheduleRequest,
    Subject,
    Teacher,
)


@dataclass(frozen=True, slots=True)
class Slot:
    index: int
    day: str
    day_index: int
    block_id: str
    block_index: int      # posición dentro del día, sólo bloques de clase
    label: str


@dataclass(slots=True)
class Lesson:
    """Una variable del CSP: una hora de una materia para un grupo."""

    uid: str
    group_id: str
    subject_id: str
    occurrence: int
    eligible_teachers: Tuple[str, ...]
    max_per_day: int
    preferred_teacher: Optional[str]
    assign_to_tutor: bool
    # Dificultad estática: menor = más difícil de colocar. Se usa como orden base
    # y como criterio de desempate del MRV.
    difficulty: float = 0.0


@dataclass
class ProblemContext:
    """Vista compilada e inmutable del problema."""

    request: ScheduleRequest
    slots: List[Slot]
    slots_by_day: List[List[int]]                 # day_index -> [slot_index...]
    day_of_slot: List[int]
    block_of_slot: List[int]
    n_blocks_per_day: int

    subjects: Dict[str, Subject]
    teachers: Dict[str, Teacher]
    groups: Dict[str, Group]
    teacher_order: List[str]
    group_order: List[str]

    teacher_slots: Dict[str, Set[int]]            # slots donde el profesor SÍ puede
    group_slots: Dict[str, Set[int]]              # slots donde el grupo SÍ recibe clase
    teacher_max_weekly: Dict[str, int]
    teacher_max_daily: Dict[str, int]

    eligible: Dict[Tuple[str, str], Tuple[str, ...]]   # (group,subject) -> profesores
    lessons: List[Lesson] = field(default_factory=list)
    required_hours: int = 0

    # ------------------------------------------------------------------ #
    def slot_index(self, day: str, block_id: str) -> Optional[int]:
        for s in self.slots:
            if s.day == day and s.block_id == block_id:
                return s.index
        return None


# --------------------------------------------------------------------------- #
# Disponibilidad
# --------------------------------------------------------------------------- #
def _teacher_available_slots(teacher: Teacher, slots: Sequence[Slot]) -> Set[int]:
    """
    Traduce la matriz de disponibilidad a un set de slots permitidos.

    Ver `docs/CONTRACT.md` §1: sin matriz => disponible siempre; blacklist => se
    quitan los listados; whitelist => sólo los listados (un día ausente queda fuera).
    """
    matrix = teacher.availability or {}
    if not matrix:
        return {s.index for s in slots}

    if teacher.availability_mode is AvailabilityMode.WHITELIST:
        allowed: Set[int] = set()
        for s in slots:
            if s.block_id in set(matrix.get(s.day, [])):
                allowed.add(s.index)
        return allowed

    # blacklist
    blocked = {(day, bid) for day, bids in matrix.items() for bid in bids}
    return {s.index for s in slots if (s.day, s.block_id) not in blocked}


# --------------------------------------------------------------------------- #
# Compilación
# --------------------------------------------------------------------------- #
def compile_problem(request: ScheduleRequest) -> ProblemContext:
    grid = request.grid
    class_blocks = grid.class_blocks

    slots: List[Slot] = []
    slots_by_day: List[List[int]] = [[] for _ in grid.days]
    idx = 0
    for d_i, day in enumerate(grid.days):
        for b_i, block in enumerate(class_blocks):
            slots.append(
                Slot(
                    index=idx,
                    day=day,
                    day_index=d_i,
                    block_id=block.id,
                    block_index=b_i,
                    label=f"{day} {block.label}",
                )
            )
            slots_by_day[d_i].append(idx)
            idx += 1

    subjects = {s.id: s for s in request.subjects}
    teachers = {t.id: t for t in request.teachers}
    groups = {g.id: g for g in request.groups}

    teacher_slots = {t.id: _teacher_available_slots(t, slots) for t in request.teachers}
    teacher_max_weekly = {t.id: t.max_weekly_hours for t in request.teachers}
    teacher_max_daily = {
        t.id: (t.max_daily_hours if t.max_daily_hours is not None else len(class_blocks))
        for t in request.teachers
    }

    # Slots hábiles por grupo (se descuentan los bloqueados por el propio grupo).
    all_slots = {s.index for s in slots}
    slot_lookup = {(s.day, s.block_id): s.index for s in slots}
    group_slots: Dict[str, Set[int]] = {}
    for g in request.groups:
        blocked = {
            slot_lookup[(b.day, b.block_id)]
            for b in g.blocked_slots
            if (b.day, b.block_id) in slot_lookup
        }
        group_slots[g.id] = all_slots - blocked

    # Elegibilidad (grupo, materia) -> profesores.
    # `fixed_teacher_id` gana sobre la lista de materias del profesor: es una orden
    # explícita de la dirección de la escuela, no una sugerencia.
    eligible: Dict[Tuple[str, str], Tuple[str, ...]] = {}
    for g in request.groups:
        for c in g.curriculum:
            if c.assign_to_tutor:
                # Se resuelve en `bind_tutor_lessons` una vez asignados los tutores.
                eligible[(g.id, c.subject_id)] = tuple()
            elif c.fixed_teacher_id:
                eligible[(g.id, c.subject_id)] = (c.fixed_teacher_id,)
            else:
                eligible[(g.id, c.subject_id)] = tuple(
                    t.id for t in request.teachers if c.subject_id in t.subject_ids
                )

    ctx = ProblemContext(
        request=request,
        slots=slots,
        slots_by_day=slots_by_day,
        day_of_slot=[s.day_index for s in slots],
        block_of_slot=[s.block_index for s in slots],
        n_blocks_per_day=len(class_blocks),
        subjects=subjects,
        teachers=teachers,
        groups=groups,
        teacher_order=[t.id for t in request.teachers],
        group_order=[g.id for g in request.groups],
        teacher_slots=teacher_slots,
        group_slots=group_slots,
        teacher_max_weekly=teacher_max_weekly,
        teacher_max_daily=teacher_max_daily,
        eligible=eligible,
    )
    _build_lessons(ctx)
    return ctx


def _build_lessons(ctx: ProblemContext) -> None:
    """Explota el plan de estudios en variables de una hora y las puntúa por dificultad."""
    lessons: List[Lesson] = []
    for gid in ctx.group_order:
        group = ctx.groups[gid]
        for c in group.curriculum:
            elig = ctx.eligible[(gid, c.subject_id)]
            for k in range(c.weekly_hours):
                lessons.append(
                    Lesson(
                        uid=f"{gid}|{c.subject_id}|{k}",
                        group_id=gid,
                        subject_id=c.subject_id,
                        occurrence=k,
                        eligible_teachers=elig,
                        max_per_day=c.max_per_day,
                        preferred_teacher=c.preferred_teacher_id,
                        assign_to_tutor=c.assign_to_tutor,
                    )
                )
    ctx.lessons = lessons
    ctx.required_hours = len(lessons)
    score_difficulty(ctx)


def score_difficulty(ctx: ProblemContext) -> None:
    """
    Dificultad ≈ tamaño del dominio inicial (slots viables × profesores elegibles).

    Es una cota superior barata que ignora las colisiones; sirve como orden base
    "primero lo más difícil", que es lo que evita el 80 % de los backtracks.
    """
    for lesson in ctx.lessons:
        g_slots = ctx.group_slots[lesson.group_id]
        total = 0
        for tid in lesson.eligible_teachers:
            total += len(g_slots & ctx.teacher_slots[tid])
        # Sin profesor todavía (assign_to_tutor) => se trata como muy restringida.
        lesson.difficulty = float(total) if lesson.eligible_teachers else 0.5


def bind_tutor_lessons(ctx: ProblemContext, tutor_by_group: Dict[str, Optional[str]]) -> None:
    """
    Segunda pasada: las materias marcadas `assign_to_tutor` sólo pueden ser impartidas
    por el tutor ya asignado al grupo. Por eso los tutores se resuelven ANTES del
    horario: aquí es donde esa decisión se convierte en una restricción dura.
    """
    for lesson in ctx.lessons:
        if not lesson.assign_to_tutor:
            continue
        tutor = tutor_by_group.get(lesson.group_id)
        lesson.eligible_teachers = (tutor,) if tutor else tuple()
        ctx.eligible[(lesson.group_id, lesson.subject_id)] = lesson.eligible_teachers
    score_difficulty(ctx)

File: assets/js/engine/domain.js
/**
 * Compilación del problema (puerto JS de `backend/app/solver/domain.py`).
 *
 * Traduce la petición normalizada a estructuras indexadas por entero. En el
 * navegador esto importa todavía más que en Python: el hot loop del backtracking
 * corre en un Web Worker sin JIT warm-up garantizado, así que se evita cualquier
 * búsqueda por string dentro de la búsqueda.
 *
 * @module engine/domain
 */

/**
 * @typedef {{index:number, day:string, dayIndex:number, blockId:string,
 *            blockIndex:number, label:string}} Slot
 * @typedef {{uid:string, groupId:string, subjectId:string, occurrence:number,
 *            eligible:string[], maxPerDay:number, preferredTeacher:?string,
 *            assignToTutor:boolean, difficulty:number}} Lesson
 */

/**
 * Traduce la matriz de disponibilidad a un Set de índices de slot permitidos.
 * Reglas en `docs/CONTRACT.md` §1.
 * @returns {Set<number>}
 */
function teacherAvailableSlots(teacher, slots) {
  const matrix = teacher.availability;
  if (!matrix || Object.keys(matrix).length === 0) {
    return new Set(slots.map((s) => s.index));
  }
  if (teacher.availability_mode === 'whitelist') {
    const allowed = new Set();
    for (const slot of slots) {
      const list = matrix[slot.day];
      if (Array.isArray(list) && list.includes(slot.blockId)) allowed.add(slot.index);
    }
    return allowed;
  }
  // blacklist
  const blocked = new Set();
  for (const [day, list] of Object.entries(matrix)) {
    for (const blockId of list || []) blocked.add(`${day}|${blockId}`);
  }
  return new Set(slots.filter((s) => !blocked.has(`${s.day}|${s.blockId}`)).map((s) => s.index));
}

/**
 * @param {object} request petición ya normalizada
 * @returns {object} contexto compilado
 */
export function compileProblem(request) {
  const classBlocks = request.grid.blocks.filter((b) => b.kind === 'class');

  /** @type {Slot[]} */
  const slots = [];
  const slotsByDay = request.grid.days.map(() => []);
  let index = 0;
  request.grid.days.forEach((day, dayIndex) => {
    classBlocks.forEach((block, blockIndex) => {
      slots.push({ index, day, dayIndex, blockId: block.id, blockIndex, label: `${day} ${block.label}` });
      slotsByDay[dayIndex].push(index);
      index += 1;
    });
  });

  const subjects = new Map(request.subjects.map((s) => [s.id, s]));
  const teachers = new Map(request.teachers.map((t) => [t.id, t]));
  const groups = new Map(request.groups.map((g) => [g.id, g]));
  const teacherOrder = request.teachers.map((t) => t.id);
  const groupOrder = request.groups.map((g) => g.id);

  const teacherSlots = new Map(request.teachers.map((t) => [t.id, teacherAvailableSlots(t, slots)]));
  const teacherMaxWeekly = new Map(request.teachers.map((t) => [t.id, t.max_weekly_hours]));
  const teacherMaxDaily = new Map(
    request.teachers.map((t) => [t.id, t.max_daily_hours ?? classBlocks.length]),
  );

  const slotLookup = new Map(slots.map((s) => [`${s.day}|${s.blockId}`, s.index]));
  const groupSlots = new Map();
  for (const group of request.groups) {
    const set = new Set(slots.map((s) => s.index));
    for (const blocked of group.blocked_slots) {
      const idx = slotLookup.get(`${blocked.day}|${blocked.block_id}`);
      if (idx !== undefined) set.delete(idx);
    }
    groupSlots.set(group.id, set);
  }

  // Elegibilidad (grupo|materia) -> profesores.
  const eligible = new Map();
  for (const group of request.groups) {
    for (const entry of group.curriculum) {
      const key = `${group.id}|${entry.subject_id}`;
      if (entry.assign_to_tutor) {
        eligible.set(key, []); // se resuelve tras asignar tutores
      } else if (entry.fixed_teacher_id) {
        eligible.set(key, [entry.fixed_teacher_id]);
      } else {
        eligible.set(
          key,
          request.teachers.filter((t) => t.subject_ids.includes(entry.subject_id)).map((t) => t.id),
        );
      }
    }
  }

  const ctx = {
    request,
    slots,
    slotsByDay,
    dayOfSlot: slots.map((s) => s.dayIndex),
    blockOfSlot: slots.map((s) => s.blockIndex),
    nBlocksPerDay: classBlocks.length,
    nDays: request.grid.days.length,
    subjects,
    teachers,
    groups,
    teacherOrder,
    groupOrder,
    teacherSlots,
    groupSlots,
    teacherMaxWeekly,
    teacherMaxDaily,
    eligible,
    /** @type {Lesson[]} */
    lessons: [],
    requiredHours: 0,
  };

  buildLessons(ctx);
  return ctx;
}

/** Explota el plan de estudios en variables de una hora. */
function buildLessons(ctx) {
  /** @type {Lesson[]} */
  const lessons = [];
  for (const gid of ctx.groupOrder) {
    for (const entry of ctx.groups.get(gid).curriculum) {
      const elig = ctx.eligible.get(`${gid}|${entry.subject_id}`) || [];
      for (let k = 0; k < entry.weekly_hours; k += 1) {
        lessons.push({
          uid: `${gid}|${entry.subject_id}|${k}`,
          groupId: gid,
          subjectId: entry.subject_id,
          occurrence: k,
          eligible: elig,
          maxPerDay: entry.max_per_day,
          preferredTeacher: entry.preferred_teacher_id,
          assignToTutor: entry.assign_to_tutor,
          difficulty: 0,
        });
      }
    }
  }
  ctx.lessons = lessons;
  ctx.requiredHours = lessons.length;
  scoreDifficulty(ctx);
}

/** Dificultad ≈ tamaño del dominio inicial (cota superior barata). */
export function scoreDifficulty(ctx) {
  for (const lesson of ctx.lessons) {
    const gSlots = ctx.groupSlots.get(lesson.groupId);
    let total = 0;
    for (const tid of lesson.eligible) {
      const tSlots = ctx.teacherSlots.get(tid);
      if (!tSlots) continue;
      for (const slot of gSlots) if (tSlots.has(slot)) total += 1;
    }
    lesson.difficulty = lesson.eligible.length ? total : 0.5;
  }
}

/**
 * Segunda pasada: las materias `assign_to_tutor` sólo puede impartirlas el tutor.
 * Por eso los tutores se resuelven ANTES del horario.
 * @param {object} ctx
 * @param {Map<string,?string>} tutorByGroup
 */
export function bindTutorLessons(ctx, tutorByGroup) {
  for (const lesson of ctx.lessons) {
    if (!lesson.assignToTutor) continue;
    const tutor = tutorByGroup.get(lesson.groupId);
    lesson.eligible = tutor ? [tutor] : [];
    ctx.eligible.set(`${lesson.groupId}|${lesson.subjectId}`, lesson.eligible);
  }
  scoreDifficulty(ctx);
}

/** Intersección de dos Sets como array ordenado (uso en precálculo, no en el hot loop). */
export function intersectSorted(a, b) {
  const out = [];
  for (const v of a) if (b.has(v)) out.push(v);
  out.sort((x, y) => x - y);
  return out;
}
